cyrilletuzi/angular-async-local-storage
# JSON Schema validation (community opinion wanted)

As mentioned in README : 

> it's client-side storage : always check the data, as it could have been forged or deleted.

So I propose to introduce some validation based on [JSON Schema](http://json-schema.org/).

To test it : 

`npm install angular-async-local-storage@beta`

The API works like this : 

```typescript
import { JSONSchema } from 'angular-async-local-storage';

const schema: JSONSchema = {
  properties: {
    expected: {
      type: 'string'
    }
  },
  required: ['expected']
};

this.localStorage.getItem<{ expected: string }>(index, { schema }).subscribe((data) => {
  /* if data is valid or null */
}, (error) => {
  /* if data or schema are invalid */
});
```

**Do NOT use in production.** The beta is working and backward-compatible, but the API may change significantly. Angular 5 is required.

My main concern is how to manage invalid data : classic behavior would be to throw an error, but as `localStorage` was never throwing before (just returning `null`), it would force people to handle the error callback.

~~For now, invalid data is just changed to `null`, but discussion is opened. I would like other opinions before going final with this feature.~~ (see next comment)

Note that not all of JSON Schema features are supported yet, just the main features to type (just follow the `JSONSchema` interface provided).
